hls-graph/src/Development/IDE/Graph/Internal/Database.hs

```diff
@@ -284,27 +288,36 @@ withRunInIO k = do
     k $ RunInIO (\aio -> runReaderT (unAIO aio) st)
 
 cleanupAsync :: IORef [Async a] -> IO ()
+-- mask to make sure we interrupt all the asyncs
 cleanupAsync ref = uninterruptibleMask_ $ do
-    asyncs <- readIORef ref
+    asyncs <- atomicModifyIORef' ref ([],)
+    -- interrupt all the asyncs without waiting
     mapM_ (\a -> throwTo (asyncThreadId a) AsyncCancelled) asyncs
-    mapM_ waitCatch asyncs
+    -- Wait until all the asyncs are done
+    -- But if it takes more than 10 seconds, log to stderr
+    unless (null asyncs) $ do
+        let loop = forever $ do
+                sleep 10
+                traceM "cleanupAsync: waiting for asyncs to finish"
```

Comment:
Do we really want to use `Debug.Trace` here?

Reply:
If you are seeing this trace, then most likely `hls-graph` is broken. 

Plus, I didn't want to thread a logger around....

Reply:
You're in IO, you could just print to stderr. It's a small difference, but I do think not having `Debug.Trace` in production code is good.

Reply:
That's a fair point